book/src/rust/rust_style_guide.md

```diff
@@ -0,0 +1,697 @@
+    Baz  => 0,
+    Quux => 1
+};
+```
+
+Over
+
+```rust
+let foo;
+match bar {
+    Baz  => {
+        foo = 0;
+    }
+    Quux => {
+        foo = 1;
+    }
+}
+```
+
+unless the conditions for initialization are too complex to fit into a simple conditional expression.
+
+##### Use type annotations for clarification; prefer explicit generics when inference fails.
+
+Prefer
+
+```rust
+s.iter().map(|x| x * 2)
+        .collect::<Vec<_>>()
+```
+
+Over
+
+```rust
+let v: Vec<_> = s.iter().map(|x| x * 2)
+                        .collect();
+```
+
+When the type of a value might be unclear to the reader of the code, consider explicitly annotating it in a let.
+
+On the other hand, when the type is unclear to the compiler, prefer to specify the type by explicit generics instantiation, which is usually more clear.
+
+##### Prefer immutable bindings.
+
+Use mut bindings to signal the span during which a value is mutated:
+
+```rust
+let mut v = Vec::new();
+// push things onto v
+let v = v;
+// use v immutably henceforth
+```
+
+#### Traits
+
+##### Defining Traits
+
+When defining a trait, use the following syntax:
+
+```rust
+/// DOCS
+trait Trait<T> {
+    /// DOCS
+    type Type1: Default;
+
+    /// DOCS
+    type Type2;
+
+    /// DOCS
+    const CONST_1: usize;
+
+    /// DOCS
+    const CONST_2: usize;
+
+    /// DOCS
+    fn required_method(&self, argument: Self::Type1) -> T;
+
+    /// DOCS
+    #[inline]
+    fn optional_method(&self) -> T {
+        Self::required_method(Self::Type1::default())
+    }
+}
+```
+
+Notice the ordering of components:
+
+1. Associated Types
+
+2. Associated Constants
+
+3. Methods
+
+Depending on the context and presentation, you can mix the ordering of required and optional methods. Also, notice the name formatting, although clippy should detect if naming differs from this pattern.
+
+##### Implementing Traits
+
+When implementing traits, use the following syntax:
+
+```rust
+impl<T> Trait for Struct<T> {
+    type Type1 = B;
+    type Type2 = C;
+
+    const CONST_1: usize = 3;
+    const CONST_2: usize = 4;
+
+    #[inline]
+    fn required_method(&self, argument: Self::Type1) -> T {
+        self.struct_method(argument).clone()
+    }
+
+    #[inline]
+    fn optional_method(&self) -> T {
+        short_cut_optimization(self)
+    }
+}
+```
+
+Notice the lack of space between implementations of the same category except for methods which always get a new line between them (like all methods). Only add space between types and constants if a comment is necessary, like in this example:
+
+```rust
+impl Configuration {
+    const SPECIAL_CONSTANT: usize = 12345;
+
+    /// In this case we have to use this constant because it's very special.
+    const ANOTHER_SPECIAL_CONSTANT: usize = 9876;
+}
+```
+
+Otherwise, it should look like:
+
+```rust
+impl Configuration {
+    const SPECIAL_CONSTANT: usize = 12345;
+    const ANOTHER_SPECIAL_CONSTANT: usize = 9876;
+}
+```
+
+#### Errors
+
+TODO: - @minikin
+
```

Comment:
Needs discussion, but we should Either:
1. always manually define custom errors.
2. Choose and then consistently use 1 library to simplify custom error definitions (such as https://crates.io/crates/snafu). 

This library is an example, not a recommendation.

We should not do a mix of the above with whatever library the developer feels like using at the time.

Reply:
https://crates.io/crates/thiserror is pretty good (and popular).

Reply:
@saibatizoku Do you remember we discussed an error guide for Rust a few weeks ago? I can't find that URL.